pkg/executor/fnStatusWatcher.go

```diff
@@ -0,0 +1,228 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package executor
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	k8sCache "k8s.io/client-go/tools/cache"
+
+	fv1 "github.com/fission/fission/pkg/apis/core/v1"
+	"github.com/fission/fission/pkg/crd"
+)
+
+const (
+	LABEL_ENV_NAME            = "envName"
+	LABEL_ENV_RESOURCEVERSION = "envResourceVersion"
+)
+
+type (
+	fnStatusWatcher struct {
+		logger        *zap.Logger
+		fissionClient *crd.FissionClient
+		k8sClient     *kubernetes.Clientset
+	}
+)
+
+func makefnStatusWatcher(logger *zap.Logger, fissionClient *crd.FissionClient,
+	k8sClientSet *kubernetes.Clientset) *fnStatusWatcher {
+
+	fnw := &fnStatusWatcher{
+		logger:        logger.Named("function_watcher"),
+		fissionClient: fissionClient,
+		k8sClient:     k8sClientSet,
+	}
+	return fnw
+}
+
+func (fnw *fnStatusWatcher) watchFunctions() {
+	lw := k8sCache.NewListWatchFromClient(fnw.fissionClient.CoreV1().RESTClient(), "functions", apiv1.NamespaceAll, fields.Everything())
+
+	_, controller := k8sCache.NewInformer(lw, &fv1.Function{}, 30*time.Second, k8sCache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			// TODO: A workaround to process items in parallel. We should use workqueue ("k8s.io/client-go/util/workqueue")
+			// and worker pattern to process items instead of moving process to another goroutine.
+			// example: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go
+			go func() {
+				fn := obj.(*fv1.Function)
+				err := fnw.updateFuncStatus(fn)
+				if err != nil {
+					fnw.logger.Error("error updating function status",
+						zap.Error(err),
+						zap.Any("function", fn))
+				}
+			}()
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			go func() {
+				fn := newObj.(*fv1.Function)
+				err := fnw.updateFuncStatus(fn)
+				if err != nil {
+					fnw.logger.Error("error updating function status",
+						zap.Error(err),
+						zap.Any("function", fn))
+				}
+			}()
+		},
+	})
+	controller.Run(make(chan struct{}))
+}
+
+func (fnw *fnStatusWatcher) updateFuncStatus(fn *fv1.Function) error {
+	// In FunctionStatus object; set the status of the package of this function
+	fnw.setFnPkgStatus(fn)
+
+	// In FunctionStatus object; set the status of the environment of this function
+	fnw.setFnEnvStatus(fn)
```

Comment:
These two are not going to work - because this method `updateFuncStatus` is called only when a function is updated. Let's imagine a package source code was updated and the build failed - then only package's create/update method is called and not function's. So what ends up happening is that package status has changed from Successful to failed - but that is not ever going to reflect here till the Function itself is updated!

Reply:
I had thought about the case u mentioned here. There could be two ways to update function status upon change in related Package or Environment:
**solution 1)** Whenever Package/Environment is updated; they should call an 'Update' notification for all the functions associated with them.(This approach is used in `build(buildCache *cache.Cache, srcpkg *fv1.Package)` method of `pkg/buildermgr/pkgwatcher.go` LN193)
**solution 2)** The Informer's `resyncPeriod` should take care of periodic status update for all the functions.(The similar approach is used in `initEnvController()` method of `pkg/executor/executortype/newdeploy/newdeploymgr.go` for updating Updating all function of the environment that changed. Here, the informer's `resyncperiod` is set to `30*time.second`) 

In current implementation; I am using **solution 2)**. This will take care updating all function's status in every 30 seconds